### ISSUE TITLE
Reconfigure Amplify at runtime (call `Amplify.configure()` multiple times)

### ISSUE THREAD

**Issue by sagar-sable** [aws-amplify/amplify-flutter#1902]
### Updated Request
The ability to call `Amplify.configure()` multiple times with new config values, or to use different config values based on region or other condition.

### Original Request
I want to configure amplify once for Online including AmplifyDataStore Plugin and AmplifyAPI Plugin and second once for offline only AmplifyDataStore Plugin.
How to remove configuration of amplify?

**Comment by ragingsquirrel3** [comment]
@sagar14791 this is not currently possible so marked this as a feature request.

**Comment by tolik505** [comment]
@ragingsquirrel3 is there any workaround besides app restart?

**Comment by ragingsquirrel3** [comment]
@sagar14791 my mistake, you can call `await Amplify.reset()` .

**Comment by sagar14791** [comment]
await Amplify.reset();
Shows Error The member 'reset' can only be used within
'package:amplify_flutter/src/amplify_impl.dart' or a test.

**Comment by sagar14791** [comment]
@ragingsquirrel3 Amplify.reset(); gives error 
"The member 'reset' can only be used within 'package:amplify_flutter/src/amplify_impl.dart' or a test."

**Comment by dnys1** [comment]
Sorry for the confusion here. To clarify, runtime re-configuration is not currently supported in the underlying iOS/Android libraries. We are tracking this as a feature request [here](https://github.com/aws-amplify/amplify-flutter/issues/1208) and will post further updates there.

**Comment by Jordan-Nelson** [comment]
I am going to re-open this issue. This was closed as a duplicate of #1208. While there are similarities between the two requests, they do seem to be distinct.

We will use this issue to track interest in the ability to re-configure Amplify at run time. That is, call Amplify.configure() multiple times.

#1208 will be used to track interest in a new service which allows pulling any configuration data from a remote source. While this configuration data *could* be an Amplify config, it could also be any other data such as feature flags.

**Comment by Jordan-Nelson** [comment]
This feature is not currently on our roadmap. If you would like to see this supported, please comment with details about your use case. Thanks.

**Comment by flodaniel** [comment]
I want to describe our use-case: 
We are offering our app in the european union and in the US. With the EU having the GDPR (data privacy law), we have duplicated our entire AWS infrastructure and are basically running a mirrored setup in eu-central-1 and us-east-1. In our flutter app the user has the option to switch between the two regions on the very first screen, which is why we would need a reconfigure() call, that allows the user to, for example, swap back from a selected us-east-1 to eu-central-1.

**Comment by sagar14791** [comment]
@Jordan-Nelson Thank you for re-opening this issue
I have One application that I am using for data collection offline and online,
for Online data Collection and sync to amplify I am using the following Configuration.

              Future<void> _configureAmplifyOnline() async {
                await Amplify.addPlugins([ 
                  AmplifyDataStore(modelProvider: ModelProvider.instance),
                  AmplifyAPI(), // to connect to amplify 
                ]);
                await Amplify.configure(amplifyconfig);
              }

for Offline data Collection and storing data Offline, I am using the following Configuration.

             Future<void> _configureAmplifyOfflnine() async {
                await Amplify.addPlugins([ 
                  AmplifyDataStore(modelProvider: ModelProvider.instance), //only to store data in oflline
                  ]);
                await Amplify.configure(amplifyconfig);
              }

now when I am online and collecting data all works fine, as soon as I enter data it goes to the datastore and gets synced to amplify,
but if all of a sudden I went offline due to a network issue, My data will get saved in the datastore and won't be synced, later when the network is there data will get synced as in my Online configuration I have **AmplifyAPI();** Plugin.
Condition 1:
If I have to send my user to areas where there is no network available, he or she will sync data(Daily Tasks) from Amplify to a Mobile device at Office or where Network is available i.e Delta sync, sync will happen once every day then they may close the application,
 after on-field, if they start the application again they will have 2 Options to select Continue in Online ( which will again sync data and it will add **AmplifyAPI();** in current configuration) or In Offlince which will not Sync data (but it will not add **AmplifyAPI();** in the configuration which will stop data sync from device to server)
then they will start the journey 

at the end of the day when they will reach the office or any time of the day they have access to the network, they should be able to Sync data 
for that, there is a button in the application for data syncing,
which should be able to change the configuration of Amplify from
 Offline             
                await Amplify.addPlugins([ 
                  AmplifyDataStore(modelProvider: ModelProvider.instance),
                ]);
to Online
                await Amplify.addPlugins([ 
                  AmplifyDataStore(modelProvider: ModelProvider.instance),
                  AmplifyAPI(), // to connect to amplify 
                ]);

So, basically I should be able to change configuration of Amplify on demand.

**Comment by sagar14791** [comment]
@ragingsquirrel3 @dnys1 @Jordan-Nelson 
My company is developing an application for a very big client using Amplify and we are stuck since last year on this Amplify Reconfiguration thing only.
Can anyone update me on this? Otherwise, we have to remove all AWS services and develop the application using other available services.

**Comment by richwcahill** [comment]
Please consider adding a runtime amplify reconfiguration.  My customers have different API endpoints, and in React + Apollo I can dynamically set those urls
Thanks!

**Comment by abdallahshaban557** [comment]
Hello everyone, thank you for all the feedback on needing this feature. We have it on our radar, and we will investigate how to enable this moving forward. We will provide an update when we have a path forward.

**Comment by krrskl** [comment]
Hello everyone, I have been interested in this function, especially my use case is to update the amplify configurations when any aws region presents a problem.

Right now the solution I see is to force the app to close and have the user start it again.

**Comment by jfacoustic** [comment]
@abdallahshaban557 Any updates on this feature?

**Comment by richwcahill** [comment]
To clarify my request....I need to modify the GQL/Appsync Endpoints based on claims made in Cognito user attributes.  Otherwise a Flutter/Amplify app can only connect with an initial configuration.  My clients who use my app have unique APIs, hence the problem with a static amplifyconfiguration.dart file.  

In an my React/Amplify app, I implement Apollo HttpLink within the client, which allows me to change the GQL URL at any time after the user has authenticated with Cognito.  Thus, allowing me to modify user access based on their attributes post authentication.
 
[https://www.apollographql.com/docs/react/api/link/apollo-link-http/](url)

However, React does not provide an easy, maintainable development path for mobile/tablet (and even desktop) like flutter does.

Is there a roadmap to implement such a feature?

**Comment by abdallahshaban557** [comment]
Hello @richwcahill  - thank you for sharing that use case. We understand this request, and have plans to enable the re-configuration of Amplify Flutter to accomplish this. We do not have an exact timeline right now - but we will share it on this issue when we do!

**Comment by Vumcam** [comment]
My use case for the runtime configuration is to enable creating a silo multi-tenant app with one frontend.
Which means that the app has a shared login page, then by using a prefix in the username I would be able to tell which backend environment the user belongs to and reconfigure the environment accordingly and grant the user access to his specific environment only.

**Comment by dJani97** [comment]
My use-case is a multi-tenant application as well. As a workaround, I'm experimenting with restarting the Flutter engine using the [restart](https://pub.dev/packages/restart) plugin.

> The plugin works by creating a new instance of the Flutter Engine. The entry point of the Dart VM is executed again, while the underlying platform specific application keeps running.

This allows me to reconfigure amplify without triggering an `AmplifyAlreadyConfiguredException`, but I'm yet to test if it causes any runtime issues after the engine is restarted and configured for a different tenant. If `Amplify.configure()` relies only on Dart code, this may work, since the Dart VM is restarted. If it calls and configures platform implementations, it will cause unexpected issues, because platform code is not restarted.

@Jordan-Nelson do you think this could work?

**Comment by anankul** [comment]
This is blocking us, in our mobile app, user can switch different countries and based on country they will go to different aws region...we can't use same aws region across countries for privacy and compliance reasons. Outside this, we also have a test app that connects to different environments that target different aws account and regions. At minimum, can Amplify.reset() method which currently is internal be made public so caller can reset and reconfigure?

**Comment by tyllark** [comment]
Hello @anankul, Amplify.reset() it is only supported for internal use within unit tests currently, but you can call it within your code by calling:
```
// ignore: invalid_use_of_visible_for_testing_member
Amplify.reset();
```
However some of our packages depend on Amplify [iOS](https://github.com/aws-amplify/amplify-swift/issues/1584)/[Android](https://github.com/aws-amplify/amplify-android/issues/560) SDKs which do not support multiple configurations, so I would not recommend using this in production code. If you can give the iOS/Android issues a thumbs up it will help us gauge interest in them. Unfortunately we are blocked on the Flutter side without a work around until these are implemented.

**Comment by isiyemijoe** [comment]
It there any update to this feature request

**Comment by mattcreaser** [comment]
Hi @isiyemijoe there is no update to share at this time, but this is something that is actively being worked on.
